[쇼핑 주문 3, 4단계] 뭉치 미션 제출합니다.

m6z1 opened the pull request:
# local.Properties 에 들어갈 값입니다!
```
base_url=http://techcourse-lv2-alb-974870821.ap-northeast-2.elb.amazonaws.com/
user_id=m6z1
user_password=password
```

# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
안녕하세요 둘리!
이번 3, 4 단계 리뷰도 잘 부탁드립니다!

<img src="https://github.com/user-attachments/assets/06e3c07f-618d-4c91-8abf-6cc09fe472fa" width="100" height="100" />

**이번 PR 의 주요한 변경 사항은 다음과 같습니다.**
1. 장바구니 화면은 쿠팡 UX 를 참고하여 페이지네이션 > 무한스크롤로 수정했습니다. 
- 무한스크롤도 제대로 작동하도록 수정했습니다!
- 장바구니 뷰모델 가독성이 떨어지는 것을 함수 분리를 통해서 조금 수정했습니다.
2. 3단계 코루틴으로 리팩토링 했습니다. 
3. 4단계 쿠폰 미션을 구현했습니다.
4. [저번 PR](https://github.com/woowacourse/android-shopping-order/pull/97) 에 달린 피드백을 반영했습니다! (코멘트도 달았습니다. 

저번 PR 에서 질문주셨던 아래의 질문에 대한 답변입니다!

> 추가로 직렬화 라이브러리 선택 이유도 다음 PR에 포함되면 좋겠습니다. 왜 reflection이 문제가 될 수 있을까요? 

리플렉션이 문제가 될 수 있는 이유는 아래와 같습니다. 
1. non-null 한 변수에 null 값이 들어갈 수 있어 null check 를 런타임에 해 kotlin 의 null-safety 한 장점이 사라집니다. (런타임에서 오류가 발생하기 때문에 디버깅이 어려워집니다)
2. 리플렉션은 런타임에 클래스 구조를 탐색하기 때문에, 객체 수가 많거나 여러 번 호출이 된다면 성능 저하가 될 수 있습니다.
3. Android에서 리플렉션 기반 클래스를 잘못 처리하면 `ClassNotFoundException`이 발생될 수 있어 안정성이 떨어집니다.

이번 쿠폰 미션에서 쿠폰이란 도메인에 고민을 많이 했습니다. 이 부분을 중점적으로 리뷰를 받고 싶습니다!

## 스크린샷

[ 쿠폰 적용 ]

[쿠폰 영상.webm](https://github.com/user-attachments/assets/5acd409d-85fb-49e7-8723-bf5c33c38471)

[장바구니 무한스크롤 수정]

[장바구니 무한스크롤 로직 수정.webm](https://github.com/user-attachments/assets/9a6929e5-8636-4de2-9e53-7372c43afaff)

m6z1 committed:
refactor: 직렬화 라이브러리 혼용 문제 해결

m6z1 committed:
test: 액티비티 두 번 실행되는 로직 수정

m6z1 committed:
refactor: 최근 본 상품 어댑터 생성 로직 수정

m6z1 committed:
refactor: gson 의존성 제거

m6z1 committed:
refactor: 사용자 인증 필요없는 retrofit 빌더 객체에 loggininterceptor 추가

m6z1 committed:
refactor: 장바구니 페이지 비어있을 때 로직 수정

m6z1 committed:
refactor: 매직넘버 상수화

m6z1 committed:
refactor: 상품 화면 item 로직 개선

m6z1 committed:
refactor: 상품 상세 화면 price MediatorLiveData 로 수정

m6z1 committed:
refactor: 불필요한 LiveData 개선

m6z1 committed:
refactor: 상품 데이터 호출 로직 코루틴으로 개선

m6z1 committed:
refactor: 매핑 함수 분리

m6z1 committed:
refactor: 장바구니 데이터 호출 로직 코루틴으로 개선

m6z1 commented:
@coderabbitai pause

m6z1 committed:
docs: 기능 요구 사항 작성

m6z1 committed:
chore: 테스트용 함수 제거

m6z1 committed:
feat: 쿠폰 도메인 구현

m6z1 committed:
feat: 쿠폰 데이터 조회 서버 통신 구현

m6z1 committed:
feat: 주문 화면 xml 구현

m6z1 committed:
refactor: 쿠폰 공통 생성자 상위 타입 프로퍼티로 수정

m6z1 committed:
chore: 도메인 패키지 변경

m6z1 committed:
feat: 배송비 도메인 객체 생성

m6z1 committed:
chore: 패키지 변경

m6z1 committed:
feat: 쿠폰 ui 모델 구현

m6z1 committed:
feat: 쿠폰 RecyclerViewAdapter 구현

m6z1 committed:
refactor: string 리소스 분리

m6z1 committed:
feat: 주문 화면 이동 로직 구현

m6z1 committed:
feat: 사용 가능한 쿠폰 출력

m6z1 committed:
feat: 쿠폰 1개 적용 로직 구현

m6z1 committed:
feat: 배송비 ui 출력 로직 구현

m6z1 committed:
docs: 기능 요구 사항 체크리스트 업데이트

m6z1 committed:
refactor: string 리소스 출력

m6z1 committed:
feat: 총 결제 금액 출력 로직 구현

m6z1 committed:
feat: 쿠폰 할인 적용 금액 로직 구현

m6z1 committed:
feat: 쿠폰 해제 시 쿠폰 할인 금액 업데이트 로직 추가

m6z1 committed:
refactor: 매직넘버 상수화

m6z1 committed:
feat: 주문 api 구현

m6z1 committed:
feat: 결제하기 버튼 클릭 로직 구현

m6z1 committed:
feat: 주문 완료 시 로직 구현

m6z1 committed:
refactor: string 리소스 분리

m6z1 committed:
feat: 사용 가능한 쿠폰 없을 경우 비어있는 뷰 구현

m6z1 committed:
refactor: 팩토리 메서드 네이밍 변경

m6z1 committed:
refactor: 불필요한 페이지네이션 객체 제거

m6z1 committed:
refactor: 장바구니 무한스크롤 로직 수정

m6z1 committed:
test: viewModel 테스트 수정

m6z1 committed:
test: OrderViewModel 테스트 추가

m6z1 committed:
fix: 이벤트 핸들링 수정

m6z1 committed:
test: ShippingFee 테스트 추가

m6z1 committed:
test: DiscountType 테스트 추가

hyemdooly commented on app/src/main/java/woowacourse/shopping/data/coupon/repository/DefaultCouponRepository.kt:
스웨거를 확인해보니 다양한 타입이 내려올 수 있더라구요 🤔 
다형성을 사용해서 파싱을 할 수도 있어요! [관련 글](https://medium.com/livefront/intro-to-polymorphism-with-kotlinx-serialization-b8f5f1cedc99)

m6z1 replied:
해당 내용 반영했습니다! a18ffd3d1a9918f660a95eb3ee8f4e348aef7454

hyemdooly commented on app/src/main/java/woowacourse/shopping/domain/order/ShippingFee.kt:
```diff
@@ -0,0 +1,10 @@
+
+@JvmInline
+value class ShippingFee(
+    val amount: Int = 3000,
```
요것도 상수화할까요?

m6z1 replied:
반영했습니다! 
363f64d42132ba74b5cba236ca16238cdaf240e4

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/order/OrderViewModel.kt:
```diff
@@ -0,0 +1,182 @@
+        _totalPrice.value = basePrice + shipping + discount
+    }
+
+    private fun calculateCouponDiscount() {
+        val couponItem = applyingCoupon.value
+        val currentPrice = price.value ?: INITIAL_AMOUNT
+
+        val discountAmount =
+            when (val origin = couponItem?.origin) {
+                is Coupon.PriceDiscount -> -origin.discount
+                is Coupon.PercentageDiscount -> -currentPrice * (origin.discountPercentage / PERCENT_BASE)
+                is Coupon.Bonus -> -(calculateBonusDiscount(origin) ?: INITIAL_AMOUNT)
+                is Coupon.FreeShipping -> -(shippingFee.value?.amount ?: INITIAL_AMOUNT)
+                null -> INITIAL_AMOUNT
+            }
```
이런 계산 로직을 쿠폰 도메인에 녹여볼 수 있을까요?

m6z1 replied:
쿠폰 할인 금액 및 쿠폰 적용 가능 여부를 도메인인 쿠폰에서 직접 판단하도록 수정했습니다.

67a037f9d4fd4cf8d956ad5975d919378f883f94

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/order/OrderViewModel.kt:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    private fun getCoupons() {
+        viewModelScope.launch {
```
왜 `viewModelScope`를 사용해야할까요?
`CoroutineScope`를 생성해서 사용하면 안되는걸까요?

m6z1 replied:
`viewModelScope` 는 뷰모델이 onCleared 생명주기가 호출될 때 함께 코루틴을 취소해주어 메모리릭의 안정성이 뛰어납니다!
CoroutineScope를 생성해서 사용해도 되지만, 직접적으로 onCleared 될 때 코루틴을 정리해주어야 하는 귀찮음이 생깁니다.

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/order/OrderViewModel.kt:
```diff
@@ -0,0 +1,182 @@
+
+    private fun getCoupons() {
+        viewModelScope.launch {
+            couponRepository
+                .getCoupons()
+                .onSuccess { coupons ->
+                    handleAvailableCoupons(coupons)
+                }.onFailure {
+                    _event.setValue(OrderEvent.GET_COUPON_FAILURE)
+                }
```
`viewModelScope`는 `Main.immeidate` 디스패쳐를 사용해요.
이 코드는 네트워크 통신인데 `withContext`로 전환하지 않아도 될까요?

m6z1 replied:
retrofit 라이브러리에서 알아서 코루틴의 디스패처를 Dispatcher.IO 로 변경시켜주는 것으로 알고 있습니다!
따라서, viewModelScope 에서 따로 디스패처를 지정해주지 않아도 문제가 없는 것으로 알고 있습니다.

hyemdooly replied:
요거 레벨인터뷰 때 custom dispatcher를 사용한다고 들었는데요!
한번 찾아보시죠 저도 다시 봐야겠어요 👀

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/order/OrderViewModel.kt:
```diff
@@ -0,0 +1,182 @@
+
+    private fun handleAvailableCoupons(coupons: List<Coupon>) {
+        val availableCoupons =
+            coupons.filter { coupon ->
+                if (coupon.isExpiration) return@filter false
+
+                when (coupon) {
+                    is Coupon.PriceDiscount -> (price.value ?: return) >= coupon.minimumAmount
+                    is Coupon.FreeShipping -> (price.value ?: return) >= coupon.minimumAmount
+                    is Coupon.PercentageDiscount -> {
+                        val now = LocalTime.now()
+                        now in coupon.availableStartTime..coupon.availableEndTime
+                    }
+
+                    is Coupon.Bonus -> {
+                        productsToOrder.any { it.quantity >= coupon.buyQuantity + coupon.getQuantity }
+                    }
+                }
+            }
```
이런 로직도 도메인 내부에서 계산하면 어떨까요?
만약 쿠폰 적용 가능 여부를 여러 곳에서 판단해야한다면, 같은 로직을 여러번 작성해주어야할 것 같아요!

m6z1 replied:
해당 내용 반영했습니다!

67a037f9d4fd4cf8d956ad5975d919378f883f94 (위 코멘트에 참조한 커밋과 동일합니다)

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/order/OrderViewModel.kt:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    fun updateApplyingCoupon(couponId: Int) {
+        val couponToApply: CouponItem = coupons.value?.find { it.id == couponId } ?: return
+        if (couponToApply.isSelected) {
+            _applyingCoupon.value = null
+            _coupons.value = coupons.value?.map { it.copy(isSelected = false) }
+            return
+        }
+        _applyingCoupon.value = couponToApply
+        updateCouponSelected(couponToApply)
```
`_applyingCoupon`은 `_coupons`에 의존적이네요!
`LiveData.map`으로 일일히 세팅하지 않고 구현할 수 있을 것 같아요.

m6z1 replied:
직접 applyingCoupon 을 업데이트 하지않고, LiveData.map 을 이용해 coupons 의 isSelected 가 true 인 값을 가지도록 로직을 수정했습니다!

f3dc12b40cc443bc9ea65041640f4350786394d7

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/product/ProductsViewModel.kt:
```diff
@@ -272,21 +275,18 @@ class ProductsViewModel(
-            result
-                .onSuccess {
+            shoppingCartRepository
+                .increaseQuantity(
+                    productItem.shoppingCartId,
+                    quantity + 1,
+                ).onSuccess {
```
보다보니 increase이지만 quantity는 increase할 양이 아니라, 업데이트할 양처럼 보이네요.
그렇다면 함수 이름을 updateQuantity로 짓는 것은 어떨까요?

m6z1 replied:
repository 에서 increase / decrease 를 구분할 필요가 없다는 것을 알게 됐습니다!
increase 와 decrease 둘 중에 한 함수를 제거 후, 네이밍을 updateQuantity 로 수정했습니다.

b78b86e91a04073b4bbdd2d965e79fd080518a1c

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/productDetail/ProductDetailViewModel.kt:
```diff
@@ -26,22 +29,37 @@ class ProductDetailViewModel(
-    private val _price: MutableLiveData<Int> = MutableLiveData()
-    val price: LiveData<Int> get() = _price
-
+    val price =
+        MediatorLiveData<Int>().apply { product.value?.product?.price }
```
이 리뷰를 참고해주세요!

m6z1 replied:
해당 내용 반영했습니다..ㅎㅎ 초기화 로직을 제대로 알고 쓰지 않아서 딱 잡혔군요 😓

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/productDetail/ProductDetailActivity.kt:
마지막으로 본 상품을 누르면 그 상품 페이지로 넘어가는 것이 요구사항이었던 것 같은데, 구현되지 않은 것 같아요!

[뭉치안넘어감.webm](https://github.com/user-attachments/assets/bdde8981-1cb5-4c8c-b860-1e01dd6af389)

m6z1 replied:
[Screen_recording_20250613_151358.webm](https://github.com/user-attachments/assets/2841ec16-a665-4f73-b957-94523513b6a9)

1, 2단계 미션 구현 당시에 문제가 생겨서 로직을 제거했었는데 깜빡했습니다 ,, ㅎㅎ ,, 해당 기능 다시 추가한 영상 같이 첨부드립니다! 🙇‍♀️

82bc636f1c85c33a16cb5f4ec9c21bb7d42a6b2d

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/shoppingCart/ShoppingCartActivity.kt:
```diff
@@ -122,37 +122,29 @@ class ShoppingCartActivity :
         val intent =
             Intent().apply {
-                putExtra("updateProducts", viewModel.hasUpdatedProducts.value)
+                putExtra("updateProducts", viewModel.hasUpdatedProducts)
```
요것도 상수화할까요?

m6z1 replied:
상수화해야 하는 것들을 반영했습니다!

363f64d42132ba74b5cba236ca16238cdaf240e4 (앞서 상수화 리뷰에 참조한 커밋과 동일합니다)

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/shoppingCartRecommend/ShoppingCartRecommendActivity.kt:
```diff
@@ -59,6 +60,7 @@ class ShoppingCartRecommendActivity :
         viewModel.recommendProducts.observe(this) { recommendItems ->
             recommendProductAdapter.submitList(recommendItems)
         }
+        viewModel.shoppingCartProductsToOrder.observe(this) { }
```
이전 리뷰에서 비슷한걸 본적 있는 것 같은데요.
이런 경우에는 꼭 이 변수가 LiveData여야하는가?를 고민해보면 좋을 것 같아요!

m6z1 replied:
LiveData 를 제거하고 List 로 변경했습니다!!
또한, 여러 라이브 데이터를 조합하는 것이 아니라면 `LiveData.map` 을 통해서 값을 업데이트하도록 수정했습니다.

9bb2d705edb011fb47fa1a623c845ed945bef295

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/shoppingCartRecommend/ShoppingCartRecommendViewModel.kt:
```diff
@@ -46,20 +48,24 @@ class ShoppingCartRecommendViewModel(
-                shoppingCartProducts = shoppingCarts.shoppingCartItems
-                getRecommendProducts()
-            }
+        viewModelScope.launch {
+            shoppingCartRepository
+                .load(0, MAX_RECENT_PRODUCT_LOAD_SIZE)
+                .onSuccess { shoppingCarts ->
+                    shoppingCartProducts = shoppingCarts.shoppingCartItems
+                    getRecommendProducts()
+                }
```
`shoppingCartProducts`를 변수로 꼭 가지고 있어야할까요?
필터링한 값들만 들고있어도 되지 않을까요? 🤔

m6z1 replied:
그러게요 ..  😓 shoppingCartProducts  를 변수로 가지지 않고 필터링만 해서 추천 상품을 업데이트 하는 식으로 개선했습니다!
꼼꼼한 리뷰 감사합니다

46851ea852352790b04a449750740ec20aa0200f

hyemdooly commented on app/src/main/java/woowacourse/shopping/data/coupon/repository/CouponRepository.kt:
사실 데이터를 받을 때 Result 형태로 매핑할 수 있어요.
CallAdapter를 사용하면 되는데요! 꽤 어려운 내용이 될 수 있을 것 같아 이 부분은 '선택'으로 남겨두겠습니다.

m6z1 replied:
요건 조금 더 공부해서 3레벨 플젝 때 적용해보겠습니다! 좋은 피드백 감사합니다 둘리 : ) 🙌

hyemdooly reviewed (changes requested):
뭉치 안녕하세요! 3,4단계 미션도 고생 많으셨습니다. 💪 
고민해보면 좋을 부분들에 대해 코멘트 남겨드렸습니다.
반영 후 재요청 부탁드려요! 마지막까지 화이팅입니다 🚀

hyemdooly commented:
[뭉치.webm](https://github.com/user-attachments/assets/19eb1f25-e40b-450f-b3a8-14d524546de2)

왜 에어포스3은 추가가 되지 않는걸까요? 404가 내려오더라구요 🤔

m6z1 committed:
refactor: 상수화

m6z1 committed:
refactor: 쿠폰 할인 금액 로직 및 사용 가능 로직 도메인으로 분리

m6z1 committed:
test: Coupon 테스트 추가

m6z1 committed:
refactor: 효과 없는 로직 제거

m6z1 committed:
feat: 상품 상세 화면 최근 본 상품 로직 구현

m6z1 committed:
refactor: repository 내 상품 수량 변경 함수 로직 개선

m6z1 committed:
refactor: 적용한 쿠폰 업데이트 로직 개선

- LiveData 의 map 함수 활용해 직접적으로 업데이트 하던 것을 개선함

m6z1 committed:
refactor: 불필요한 변수 제거

m6z1 committed:
refactor: MediatorLiveData > LiveData.map 으로 로직 수정

m6z1 committed:
refactor: 주문할 장바구니 목록 LiveData 제거

m6z1 committed:
fix: 추천 상품 로직 수정

- 전) 로컬에 저장된 최근 본 상품들 중에서 가장 최근 본 상품의 카테고리를 기준으로 데이터를 가져옴
- 후) 서버에서 가장 최근 본 상품의 카테고리를 기준으로 데이터를 가져옴

m6z1 committed:
refactor: 쿠폰 서버 응답값 다형성 이용해 로직 개선

m6z1 reviewed (commented):
안녕하세요 둘리!
레벨 2 마지막 리뷰 요청이네요 ㅎㅎ
레벨 1 부터 레벨 2까지 꼼꼼하고 도움되는 리뷰 정말 감사합니다 : )

피드백 달아주신 것에 대해서는 모두 반영했습니다!

추가로 에어포스 3 이 404 에러가 뜬 이유는 추천 상품 로직의 문제였습니다.

기존 추천 상품의 데이터를 불러오는 로직은 아래와 같습니다.
1. 최근 본 상품들 로컬 db 에 저장
2. 가장 최근에 본 상품 카테고리 기준으로 최근 본 상품들에서 상품 가져옴
3. 추천 상품으로 보여줌
-> 이 때, 에어포스 3 이 저장되어 보이게 됐습니다. 하지만, 서버에는 에어포스 3 이 제거된 상태라 404 에러가 떴던 것 같습니다. 😓

현재는 서버에서 상품들을 가져와 추천 상품을 보여주는 방식으로 해결했습니다!

레벨 1, 2 동안 정말 감사했습니다 둘리 ㅎ ㅎ 🦖

hyemdooly reviewed (approved):
뭉치 안녕하세요! 쇼핑 주문 3, 4단계 미션도 고생 많으셨습니다. 💯 
레벨2가 정신 없었을텐데 결국 끝까지 완주하셨네요! 많은 성장을 했다고 뭉치 스스로도 느끼실거에요.
레벨 1, 2에서 배운 내용들이 레벨 3부터 있을 팀프로젝트에서도 많은 도움이 되었으면 좋겠습니다.
앞으로 남은 우테코 여정도 응원할게요~! 🚀
